File: Benchmarks/4.4-manejo-peticiones/prueba-2/javascript/worker.js
const WEBSOCKET_SERVER = "ws://localhost:5002";

self.onmessage = async function (e) {
    const data = e.data;

    if (data.type === "do_analisis") {
        try {
            const { id, num_requests, delay } = data;

            const results = [];
            const individualTimes = [];

            const ws = new WebSocket(WEBSOCKET_SERVER);
            await new Promise((resolve) => {
                ws.onopen = resolve;
            });

            const pendingRequests = new Map();
            let requestId = 0;

            ws.onmessage = (event) => {
                const parsed = JSON.parse(event.data);
                const respId = parsed.id;

                if (pendingRequests.has(respId)) {
                    const { resolve, start } = pendingRequests.get(respId);
                    const elapsed = performance.now() - start;

                    individualTimes[respId] = elapsed;
                    results[respId] = parsed;

                    resolve();
                    pendingRequests.delete(respId);
                }
            };

            const sendRequestWithTiming = async (currentDelay) => {
                const currentId = requestId++;
                const payload = JSON.stringify({ delay: currentDelay, id: currentId });

                return new Promise((resolve) => {
                    pendingRequests.set(currentId, {
                        resolve,
                        start: performance.now(),
                    });
                    ws.send(payload);
                });
            };


            const fetchPromises = Array.from({ length: num_requests }, () =>
                sendRequestWithTiming(delay)
            );

            const start_time = performance.now();
            await Promise.all(fetchPromises);
            const total_time = performance.now() - start_time;

            ws.close();

            const avg_time =
                individualTimes.length > 0
                    ? individualTimes.reduce((acc, t) => acc + t, 0) /
                    individualTimes.length
                    : 0;

            let last_value = null;
            for (let i = results.length - 1; i >= 0; i--) {
                const r = results[i];
                if (r && r.data && Array.isArray(r.data)) {
                    last_value = r.data[r.data.length - 1].value;
                    break;
                }
            }

            const result = {
                average_time_ms: avg_time,
                total_time_ms: total_time,
                total_requests: num_requests,
                last_value
            };

            self.postMessage({ id, json: JSON.stringify(result) });
        } catch (error) {
            self.postMessage({
                id: data.id,
                error: `Error in worker: ${error.message}`,
            });
        }
    }
};

File: Benchmarks/4.4-manejo-peticiones/prueba-2/python/worker.py
from pyscript import sync, display, document
import asyncio
import time
import json
import js  # type: ignore

PYTHON_WS_SERVER = "ws://localhost:5001"


async def do_analisis():
    try:
        num_requests = int(document.getElementById(
            "num-requests-pyscript").value)
        delay = int(document.getElementById("request-delay-pyscript").value)

        start_time = time.perf_counter()
        result_data = await websocket_benchmark(num_requests, delay, PYTHON_WS_SERVER)
        total_time = (time.perf_counter() - start_time) * 1000

        individual_times = result_data["individual_times"]
        responses = result_data["responses"]

        avg_time = sum(individual_times) / \
            len(individual_times) if individual_times else 0

        last_value = None
        for r in reversed(responses):
            if r and "data" in r and isinstance(r["data"], list):
                last_value = r["data"][-1]["value"]
                break

        result = {
            "average_time_ms": avg_time,
            "total_time_ms": total_time,
            "total_requests": num_requests,
            "last_value": last_value
        }

        return json.dumps(result)
    except Exception as e:
        display(f"Error: {e}", target="pyscript-output")


async def websocket_benchmark(num_requests, delay, server_url):
    socket = js.WebSocket.new(server_url)

    while socket.readyState != js.WebSocket.OPEN:
        await asyncio.sleep(0.01)

    pending = {}
    responses = []
    individual_times = []

    def on_message(event):
        data = json.loads(event.data)
        req_id = data.get("id")
        if req_id in pending:
            start_time = pending.pop(req_id)
            elapsed = (time.perf_counter() - start_time) * 1000
            individual_times.append(elapsed)
            responses.append(data)

    socket.onmessage = on_message

    for i in range(num_requests):
        pending[i] = time.perf_counter()
        socket.send(json.dumps({"delay": delay, "id": i}))

    while len(responses) < num_requests:
        await asyncio.sleep(0.01)

    socket.close()
    return {"responses": responses, "individual_times": individual_times}


sync.do_analisis = do_analisis
